### ISSUE TITLE
Keras `image_dataset_from_directory` shuffles labels

### ISSUE THREAD

**Issue by CaptainDario** [tensorflow/tensorflow#50124]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.5.0
- Python version: 3.8
- CUDA version: 
```
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Sun_Feb_14_22:08:44_Pacific_Standard_Time_2021
Cuda compilation tools, release 11.2, V11.2.152
Build cuda_11.2.r11.2/compiler.29618528_0
```
- cuDNN version: `cudnn-11.2-windows-x64-v8.1.1.33`
- GPU model and memory: RTX 2070 Super

**Describe the current behavior**
I have a folder structure with ~6500 different classes.
The structure is the following:
```
etlcdb
|->  0000
|       -> 0.jpg
|       -> 1.jpg
|       ...
|-> 0001
|       -> 0.jpg
|       -> 1.jpg
|       ...
|   .
|   .
|   .
|-> 6542
|       -> 0.jpg
|       -> 1.jpg
|       ...
```
I am using `tf.keras.preprocessing.image_dataset_from_directory` to create a `tf.data.dataset` from this folder structure.
Like this:
```
#batch size
bs=512
# class names
classes = ["%04d" % i for i in range(len(labels))]

train_dataset = tf.keras.preprocessing.image_dataset_from_directory(
    directory=r'F:\data_sets\etlcdb',
    labels="inferred",
    label_mode="categorical",
    class_names=classes,
    color_mode="grayscale",
    batch_size=bs,
    image_size=(64, 64),
    validation_split=0.15,
    subset="training",
    seed=123
)
```
Output:
```
Found 6731099 files belonging to 6543 classes.
Using 5721435 files for training.
```
Afterwards I define some processing:
```
train = train_dataset.map(
    lambda x, y : (tf.cast(x, tf.float16), (tf.cast(y, tf.float16))),
    num_parallel_calls=tf.data.AUTOTUNE
)
train = train.cache(r"F:\data_sets\etlcdb_cache\cache_train")
#train = train.shuffle(buffer_size=bs*3)
train = train.prefetch(buffer_size=tf.data.AUTOTUNE)
```
Finally I train the network with `fit()`.<br/><br/>
When than predictions are made the labels do not match anymore.
With a variable `labels` which contains the class labels matching the folder structure (attached as a .txt file) I execute following code:
```
sample = tf.keras.preprocessing.image.load_img(
    path= r"F:\data_sets\etlcdb\0002\3.jpg",
    color_mode="grayscale"
)
sample = tf.keras.preprocessing.image.img_to_array(sample)
sample = sample.reshape((1, 64, 64, 1))

prediction = f16_model.predict(sample)
```
Result (the text on top of the image is the prediction and its "accuracy"):
![image](https://user-images.githubusercontent.com/51273483/121015823-b70ef080-c79b-11eb-984f-4d61e49e5e7f.png)
Because the CNN does detect all 2's as 惹 I am certain that the labels somehow get mixed up.

**Describe the expected behavior**
The labels inferred by the folder structure should line-up with the output tensor 
from the last layer of the CNN.

**Standalone code to reproduce the issue**
The notebook where this occurred can be found [here](https://github.com/CaptainDario/DaKanji-ML/blob/feature-recognize_more_characters/single_kanji_cnn/single_kanji_cnn_training.ipynb).
But the appropriate folder structure needs to be created.

[labels.txt](https://github.com/tensorflow/tensorflow/files/6608772/labels.txt)

**Comment by CaptainDario** [comment]
Is the assumption that the inferred labels should have the following structure wrong?

| folder name | one hot vector|
| ------------- | --- |
| 0000 | [1, 0, 0, ..., 0] |
| 0001 | [0, 1, 0, ..., 0] |
| 0002 | [0, 0, 1, ..., 0] |
| 6542 | [0, 0, 0, ..., 1] |

**Comment by CaptainDario** [comment]
After further investigation it seems that `class_names=classes` is being ignored.
Keras a mapping of the following structure:

| folder name | inferred one hot vector |
| :---: | :---: |
| 0000 | [1, 0, 0, 0, 0, ..., 0] |
| 0001 | [0, 1, 0, 0, 0, ..., 0] |
| 0010 | [0, 0, 1, 0, 0, ..., 0] |
| 0100 | [0, 0, 0, 1, 0, ..., 0] |
| 1000 | [0, 0, 0, 0, 1, ..., 0] |

@UsharaniPagadala should setting `class_names=["0000", "0001", "0002", "0003", "0004", ....]` not force the order I want (like the table in the comment above)?

**Comment by CaptainDario** [comment]
If I order the labels like this to match keras inferred labels everything works fine.

``` python
ls = labels_1 + labels_2
# order the labels
indexs = sorted([str(i) for i in range(0, len(ls))])
ordered_labels = [ls[int(i)] for i in indexs]
```

Do I not understand the `class_names` argument correctly or is this a bug?

**Comment by CaptainDario** [comment]
@UsharaniPagadala it seems to be `tf 2.5.0` related. I ran the same notebook on `tf 2.4.0` and there the ordering is correct.

**Comment by jvishnuvardhan** [comment]
@CaptainDario Can you please check with `tf-nightly` and let us know whether the issue persists. Thanks!

**Comment by CaptainDario** [comment]
I will try with `tf-nightly` in the coming days and report back here.

**Comment by CaptainDario** [comment]
Tried it an the problem persists with `tf-nightly==2.6.0.dev20210616`.

**Comment by jvishnuvardhan** [comment]
@CaptainDario If you can create a simple standalone code to reproduce the issue, then I can test it myself. Currently I am not sure what is the root-cause. Can you please create a colab gist or jupyter notebook that we can use to reproduce? Thanks!

**Comment by CaptainDario** [comment]
@jvishnuvardhan I will try to create a standalone notebook in the coming days.

**Comment by CaptainDario** [comment]
@jvishnuvardhan is there away to give you my dataset directly?
I cannot reproduce the issue with anythin else and caching the dataset takes to much space on colab.

**Comment by jvishnuvardhan** [comment]
@CaptainDario If the issue is with your data (as you can not reproduce with other data), then this is not a TF issue. TF repository focus on bugs/performance related issues. We cannot debug your data. Thanks!

**Comment by CaptainDario** [comment]
How can wrongly inferred labels be caused by my data?
My dataset has 6000+ classes and is highly imbalanced and something of that causes the labels to be mixed up.

However if this is too much of an edge case feel free to close this.

**Comment by teeters** [comment]
I seem to be having the same issue. I have a model that uses inferred labels for a large image dataset, and I mapped them to numeric ids using the class_names argument. It was all working properly until a few days ago.

**Comment by CaptainDario** [comment]
@teeters could you try removin a part of the samples and see if the issue persists? For me it was solved when i removed 50% of the samples of every class.

**Comment by teeters** [comment]
So, I haven't tried removing data, but I did try visualizing some of the images using this code:

```
import matplotlib.pyplot as plt 

plt.figure(figsize=(10, 10))
for images, labels in train_dataset.take(1):
  for i in range(9):
    ax = plt.subplot(3, 3, i + 1)
    plt.imshow(images[i].numpy().astype("uint8"))
    label = tf.argmax(labels[i]).numpy()
    plt.title(str(label)+' '+class_names[label])
    plt.axis("off")
```

(`class_names` is a dict that maps the string class names to their desired numeric ids). The images show up with the correct ids and string names. So maybe the shuffling is happening somewhere else in the training process.

**Comment by teeters** [comment]
> After further investigation it seems that `class_names=classes` is being ignored.
> Keras infers the label order like the following:
> 
> folder name	inferred one hot vector
> 0000	[1, 0, 0, 0, 0, ..., 0]
> 0001	[0, 1, 0, 0, 0, ..., 0]
> 0010	[0, 0, 1, 0, 0, ..., 0]
> 0100	[0, 0, 0, 1, 0, ..., 0]
> 1000	[0, 0, 0, 0, 1, ..., 0]
> @UsharaniPagadala should setting `class_names=["0000", "0001", "0002", "0003", "0004", ....]` not force the order I want?
> 
> **Desired order:**
> 
> folder name	one hot vector
> 0000	[1, 0, 0, ..., 0]
> 0001	[0, 1, 0, ..., 0]
> 0002	[0, 0, 1, ..., 0]
> ...	...
> 6542	[0, 0, 0, ..., 1]

@CaptainDario how did you view the one-hot vectors in this post?

**Comment by CaptainDario** [comment]
> 
> 
> > After further investigation it seems that `class_names=classes` is being ignored.
> > Keras infers the label order like the following:
> > folder name	inferred one hot vector
> > 0000	[1, 0, 0, 0, 0, ..., 0]
> > 0001	[0, 1, 0, 0, 0, ..., 0]
> > 0010	[0, 0, 1, 0, 0, ..., 0]
> > 0100	[0, 0, 0, 1, 0, ..., 0]
> > 1000	[0, 0, 0, 0, 1, ..., 0]
> > @UsharaniPagadala should setting `class_names=["0000", "0001", "0002", "0003", "0004", ....]` not force the order I want?
> > **Desired order:**
> > folder name	one hot vector
> > 0000	[1, 0, 0, ..., 0]
> > 0001	[0, 1, 0, ..., 0]
> > 0002	[0, 0, 1, ..., 0]
> > ...	...
> > 6542	[0, 0, 0, ..., 1]
> 
> @CaptainDario how did you view the one-hot vectors in this post?

@teeters  I trained a model on my data and gave it examples with the specific label.
And allways when I gave a sample from class "0010" it returned `[0, 0, 1, ...]`. Therefore I am sure that it actually learned the right thing just with the wrong label...

**Comment by sachinprasadhs** [comment]
`tf.keras.preprocessing.image_dataset_from_directory` will be deprecated from Tensorflow 2.9 version, prefer loading data with [tf.keras.utils.image_dataset_from_directory](https://www.tensorflow.org/api_docs/python/tf/keras/utils/image_dataset_from_directory), and then transforming the output [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) with preprocessing layers.

**Comment by CaptainDario** [comment]
I tried this function and it seems to work, Thanks!
